File: packages/java/anip-graphql/src/main/java/dev/anip/graphql/SchemaBuilder.java
package dev.anip.graphql;

import dev.anip.core.ANIPError;
import dev.anip.core.CapabilityDeclaration;
import dev.anip.core.CapabilityInput;
import dev.anip.core.CapabilityRequirement;
import dev.anip.core.Constants;
import dev.anip.core.DelegationToken;
import dev.anip.service.ANIPService;
import dev.anip.service.InvokeOpts;

import graphql.schema.DataFetcher;
import graphql.schema.DataFetchingEnvironment;
import graphql.schema.GraphQLArgument;
import graphql.schema.GraphQLFieldDefinition;
import graphql.schema.GraphQLInputType;
import graphql.schema.GraphQLNonNull;
import graphql.schema.GraphQLObjectType;
import graphql.schema.GraphQLOutputType;
import graphql.schema.GraphQLSchema;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import static graphql.Scalars.*;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
import static graphql.schema.GraphQLObjectType.newObject;
import static graphql.schema.GraphQLNonNull.nonNull;

/**
 * Builds a graphql-java schema at runtime from ANIP capabilities.
 * Query for read, Mutation for write. CamelCase field names.
 */
public class SchemaBuilder {

    private SchemaBuilder() {}

    /**
     * Maps an ANIP type name to a GraphQL output type.
     */
    public static GraphQLOutputType anipTypeToGraphQL(String anipType) {
        return switch (anipType) {
            case "string", "date", "airport_code" -> GraphQLString;
            case "integer" -> GraphQLInt;
            case "number" -> GraphQLFloat;
            case "boolean" -> GraphQLBoolean;
            default -> ExtendedScalars.JSON;
        };
    }

    /**
     * Maps an ANIP type name to a GraphQL type name string (for SDL).
     */
    public static String anipTypeToGraphQLName(String anipType) {
        return switch (anipType) {
            case "string", "date", "airport_code" -> "String";
            case "integer" -> "Int";
            case "number" -> "Float";
            case "boolean" -> "Boolean";
            default -> "JSON";
        };
    }

    /**
     * Builds an executable GraphQLSchema from the service's capabilities.
     */
    @SuppressWarnings("unchecked")
    public static SchemaResult buildSchema(ANIPService service) {
        Map<String, Object> manifest = (Map<String, Object>) service.getManifest();
        Map<String, Object> capabilities = (Map<String, Object>) manifest.get("capabilities");

        // Sorted for deterministic output.
        List<String> capNames = new ArrayList<>(new TreeMap<>(capabilities).keySet());

        // Shared types.
        GraphQLObjectType financialCostType = newObject()
                .name("FinancialCost")
                .field(newFieldDefinition().name("amount").type(GraphQLFloat))
                .field(newFieldDefinition().name("currency").type(GraphQLString))
                .build();

        GraphQLObjectType costActualType = newObject()
                .name("CostActual")
                .field(newFieldDefinition().name("financial").type(financialCostType))
                .field(newFieldDefinition().name("varianceFromEstimate").type(GraphQLString))
                .build();

        GraphQLObjectType resolutionType = newObject()
                .name("Resolution")
                .field(newFieldDefinition().name("action").type(nonNull(GraphQLString)))
                .field(newFieldDefinition().name("recoveryClass").type(nonNull(GraphQLString)))
                .field(newFieldDefinition().name("requires").type(GraphQLString))
                .field(newFieldDefinition().name("grantableBy").type(GraphQLString))
                .build();

        GraphQLObjectType anipFailureType = newObject()
                .name("ANIPFailure")
                .field(newFieldDefinition().name("type").type(nonNull(GraphQLString)))
                .field(newFieldDefinition().name("detail").type(nonNull(GraphQLString)))
                .field(newFieldDefinition().name("resolution").type(resolutionType))
                .field(newFieldDefinition().name("retry").type(nonNull(GraphQLBoolean)))
                .build();

        List<GraphQLFieldDefinition> queryFields = new ArrayList<>();
        List<GraphQLFieldDefinition> mutationFields = new ArrayList<>();

        for (String name : capNames) {
            CapabilityDeclaration decl = service.getCapabilityDeclaration(name);
            if (decl == null) continue;

            String pascal = GraphQLResponseMapper.toPascalCase(name);
            String camel = GraphQLResponseMapper.toCamelCase(name);

            GraphQLObjectType resultType = newObject()
                    .name(pascal + "Result")
                    .field(newFieldDefinition().name("success").type(nonNull(GraphQLBoolean)))
                    .field(newFieldDefinition().name("result").type(ExtendedScalars.JSON))
                    .field(newFieldDefinition().name("costActual").type(costActualType))
                    .field(newFieldDefinition().name("failure").type(anipFailureType))
                    .build();

            // Build args.
            List<GraphQLArgument> args = new ArrayList<>();
            if (decl.getInputs() != null) {
                for (CapabilityInput inp : decl.getInputs()) {
                    String argName = GraphQLResponseMapper.toCamelCase(inp.getName());
                    GraphQLOutputType gqlType = anipTypeToGraphQL(inp.getType());
                    GraphQLInputType inputType = inp.isRequired()
                            ? GraphQLNonNull.nonNull((GraphQLInputType) gqlType)
                            : (GraphQLInputType) gqlType;
                    args.add(GraphQLArgument.newArgument()
                            .name(argName)
                            .type(inputType)
                            .build());
                }
            }

            DataFetcher<?> fetcher = makeResolver(service, name);

            GraphQLFieldDefinition.Builder fieldBuilder = newFieldDefinition()
                    .name(camel)
                    .type(nonNull(resultType))
                    .dataFetcher(fetcher);
            for (GraphQLArgument arg : args) {
                fieldBuilder.argument(arg);
            }

            GraphQLFieldDefinition field = fieldBuilder.build();

            String seType = decl.getSideEffect() != null ? decl.getSideEffect().getType() : "";
            if (seType.isEmpty() || "read".equals(seType)) {
                queryFields.add(field);
            } else {
                mutationFields.add(field);
            }
        }

        // Build schema.
        GraphQLSchema.Builder schemaBuilder = GraphQLSchema.newSchema();

        if (!queryFields.isEmpty()) {
            GraphQLObjectType.Builder queryBuilder = newObject().name("Query");
            for (GraphQLFieldDefinition f : queryFields) {
                queryBuilder.field(f);
            }
            schemaBuilder.query(queryBuilder.build());
        } else {
            // graphql-java requires a Query type; add a dummy.
            schemaBuilder.query(newObject().name("Query")
                    .field(newFieldDefinition().name("_empty").type(GraphQLString)
                            .dataFetcher(env -> null))
                    .build());
        }

        if (!mutationFields.isEmpty()) {
            GraphQLObjectType.Builder mutationBuilder = newObject().name("Mutation");
            for (GraphQLFieldDefinition f : mutationFields) {
                mutationBuilder.field(f);
            }
            schemaBuilder.mutation(mutationBuilder.build());
        }

        GraphQLSchema schema = schemaBuilder.build();
        String sdl = generateSDL(service, capNames);

        return new SchemaResult(schema, sdl);
    }

    /**
     * Creates a data fetcher for a given capability.
     */
    private static DataFetcher<Map<String, Object>> makeResolver(ANIPService service,
                                                                    String capName) {
        return (DataFetchingEnvironment env) -> {
            // Extract auth header from context.
            String authHeader = env.getGraphQlContext().get("authHeader");
            String bearer = extractBearer(authHeader);

            if (bearer == null || bearer.isEmpty()) {
                return GraphQLResponseMapper.buildGraphQLResponse(Map.of(
                        "success", false,
                        "failure", Map.of(
                                "type", Constants.FAILURE_AUTH_REQUIRED,
                                "detail", "Authorization header required",
                                "resolution", Map.of("action", "provide_credentials", "recovery_class", Constants.recoveryClassForAction("provide_credentials")),
                                "retry", true
                        )
                ));
            }

            DelegationToken token;
            try {
                token = GraphQLAuthBridge.resolveAuth(bearer, service, capName);
            } catch (ANIPError e) {
                Map<String, Object> failure = new LinkedHashMap<>();
                failure.put("type", e.getErrorType());
                failure.put("detail", e.getDetail());
                failure.put("retry", e.isRetry());
                if (e.getResolution() != null) {
                    failure.put("resolution", Map.of(
                            "action", e.getResolution().getAction(),
                            "recovery_class", e.getResolution().getRecoveryClass() != null
                                    ? e.getResolution().getRecoveryClass()
                                    : Constants.recoveryClassForAction(e.getResolution().getAction())
                    ));
                }
                return GraphQLResponseMapper.buildGraphQLResponse(Map.of(
                        "success", false,
                        "failure", failure
                ));
            } catch (Exception e) {
                return GraphQLResponseMapper.buildGraphQLResponse(Map.of(
                        "success", false,
                        "failure", Map.of(
                                "type", Constants.FAILURE_INTERNAL_ERROR,
                                "detail", "Authentication failed",
                                "retry", false
                        )
                ));
            }

            // Convert camelCase args to snake_case.
            Map<String, Object> snakeArgs = new LinkedHashMap<>();
            for (Map.Entry<String, Object> entry : env.getArguments().entrySet()) {
                snakeArgs.put(GraphQLResponseMapper.toSnakeCase(entry.getKey()), entry.getValue());
            }

            Map<String, Object> result = service.invoke(capName, token, snakeArgs,
                    new InvokeOpts());

            return GraphQLResponseMapper.buildGraphQLResponse(result);
        };
    }

    private static String extractBearer(String authHeader) {
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            return authHeader.substring(7).trim();
        }
        return null;
    }

    /**
     * Generates SDL text including custom @anip* directives.
     */
    private static String generateSDL(ANIPService service, List<String> capNames) {
        List<String> lines = new ArrayList<>();

        // Directives.
        lines.add("directive @anipSideEffect(type: String!, rollbackWindow: String) on FIELD_DEFINITION");
        lines.add("directive @anipCost(certainty: String!, currency: String, rangeMin: Float, rangeMax: Float) on FIELD_DEFINITION");
        lines.add("directive @anipRequires(capabilities: [String!]!) on FIELD_DEFINITION");
        lines.add("directive @anipScope(scopes: [String!]!) on FIELD_DEFINITION");
        lines.add("");
        lines.add("scalar JSON");
        lines.add("");

        // Shared types.
        lines.add("type CostActual { financial: FinancialCost, varianceFromEstimate: String }");
        lines.add("type FinancialCost { amount: Float, currency: String }");
        lines.add("type ANIPFailure { type: String!, detail: String!, resolution: Resolution, retry: Boolean! }");
        lines.add("type Resolution { action: String!, recoveryClass: String!, requires: String, grantableBy: String }");
        lines.add("type RestrictedCapability { capability: String!, reason: String!, reasonType: String!, grantableBy: String!, unmetTokenRequirements: [String!]!, resolutionHint: String }");
        lines.add("type DeniedCapability { capability: String!, reason: String!, reasonType: String! }");
        lines.add("");

        List<String> queries = new ArrayList<>();
        List<String> mutations = new ArrayList<>();

        for (String name : capNames) {
            CapabilityDeclaration decl = service.getCapabilityDeclaration(name);
            if (decl == null) continue;

            String pascal = GraphQLResponseMapper.toPascalCase(name);
            String camel = GraphQLResponseMapper.toCamelCase(name);

            // Result type.
            lines.add("type " + pascal + "Result { success: Boolean!, result: JSON, costActual: CostActual, failure: ANIPFailure }");

            String argsStr = buildSDLArgs(decl);
            String dirStr = buildSDLDirectives(decl);

            String fieldLine = "  " + camel + argsStr + ": " + pascal + "Result! " + dirStr;

            String seType = decl.getSideEffect() != null ? decl.getSideEffect().getType() : "";
            if (seType.isEmpty() || "read".equals(seType)) {
                queries.add(fieldLine);
            } else {
                mutations.add(fieldLine);
            }
        }

        lines.add("");
        if (!queries.isEmpty()) {
            lines.add("type Query {");
            lines.addAll(queries);
            lines.add("}");
        }
        if (!mutations.isEmpty()) {
            lines.add("type Mutation {");
            lines.addAll(mutations);
            lines.add("}");
        }

        return String.join("\n", lines);
    }

    private static String buildSDLArgs(CapabilityDeclaration decl) {
        if (decl.getInputs() == null || decl.getInputs().isEmpty()) {
            return "";
        }
        List<String> args = new ArrayList<>();
        for (CapabilityInput inp : decl.getInputs()) {
            String argName = GraphQLResponseMapper.toCamelCase(inp.getName());
            String gqlType = anipTypeToGraphQLName(inp.getType());
            if (inp.isRequired()) {
                gqlType += "!";
            }
            args.add(argName + ": " + gqlType);
        }
        return "(" + String.join(", ", args) + ")";
    }

    @SuppressWarnings("unchecked")
    private static String buildSDLDirectives(CapabilityDeclaration decl) {
        List<String> parts = new ArrayList<>();

        // @anipSideEffect
        String seType = decl.getSideEffect() != null ? decl.getSideEffect().getType() : "";
        if (seType.isEmpty()) seType = "read";
        StringBuilder seDir = new StringBuilder("@anipSideEffect(type: \"" + seType + "\"");
        if (decl.getSideEffect() != null && decl.getSideEffect().getRollbackWindow() != null
                && !decl.getSideEffect().getRollbackWindow().isEmpty()) {
            seDir.append(", rollbackWindow: \"").append(decl.getSideEffect().getRollbackWindow()).append("\"");
        }
        seDir.append(")");
        parts.add(seDir.toString());

        // @anipCost
        if (decl.getCost() != null) {
            String certainty = decl.getCost().getCertainty();
            if (certainty == null || certainty.isEmpty()) certainty = "estimate";
            StringBuilder costDir = new StringBuilder("@anipCost(certainty: \"" + certainty + "\"");
            if (decl.getCost().getFinancial() != null) {
                String currency = decl.getCost().getFinancial().getCurrency();
                if (currency != null) {
                    costDir.append(", currency: \"").append(currency).append("\"");
                }
                Double rangeMin = decl.getCost().getFinancial().getRangeMin();
                Double rangeMax = decl.getCost().getFinancial().getRangeMax();
                if (rangeMin != null) {
                    costDir.append(", rangeMin: ").append(rangeMin);
                }
                if (rangeMax != null) {
                    costDir.append(", rangeMax: ").append(rangeMax);
                }
            }
            costDir.append(")");
            parts.add(costDir.toString());
        }

        // @anipRequires
        if (decl.getRequires() != null && !decl.getRequires().isEmpty()) {
            List<String> capNames = new ArrayList<>();
            for (CapabilityRequirement r : decl.getRequires()) {
                capNames.add("\"" + r.getCapability() + "\"");
            }
            parts.add("@anipRequires(capabilities: [" + String.join(", ", capNames) + "])");
        }

        // @anipScope
        if (decl.getMinimumScope() != null && !decl.getMinimumScope().isEmpty()) {
            List<String> scopeVals = new ArrayList<>();
            for (String s : decl.getMinimumScope()) {
                scopeVals.add("\"" + s + "\"");
            }
            parts.add("@anipScope(scopes: [" + String.join(", ", scopeVals) + "])");
        }

        return String.join(" ", parts);
    }

    /**
     * Result of building the schema: executable schema + SDL text.
     */
    public record SchemaResult(GraphQLSchema schema, String sdl) {}
}

File: packages/csharp/src/Anip.GraphQL/SchemaBuilder.cs
using System.Text;
using System.Text.Json;
using Anip.Core;
using Anip.Service;
using GraphQL;
using GraphQL.Execution;
using GraphQL.Types;

namespace Anip.GraphQL;

/// <summary>
/// Builds a graphql-dotnet schema at runtime from ANIP capabilities.
/// Query for read, Mutation for write. CamelCase field names.
/// </summary>
public static class SchemaBuilder
{
    /// <summary>
    /// Maps an ANIP type name to the corresponding GraphQL type name string (for SDL).
    /// </summary>
    public static string AnipTypeToGraphQLName(string anipType) => anipType switch
    {
        "string" or "date" or "airport_code" => "String",
        "integer" => "Int",
        "number" => "Float",
        "boolean" => "Boolean",
        _ => "JSON",
    };

    /// <summary>
    /// Maps an ANIP type name to a graphql-dotnet scalar graph type.
    /// </summary>
    private static IGraphType AnipTypeToGraphQL(string anipType) => anipType switch
    {
        "string" or "date" or "airport_code" => new StringGraphType(),
        "integer" => new IntGraphType(),
        "number" => new FloatGraphType(),
        "boolean" => new BooleanGraphType(),
        _ => new JsonGraphType(),
    };

    /// <summary>
    /// Builds an executable ISchema and SDL text from the service's capabilities.
    /// </summary>
    public static SchemaResult BuildSchema(AnipService service)
    {
        var manifest = service.GetManifest();
        var capabilities = manifest.Capabilities;

        // Sorted for deterministic output.
        var capNames = capabilities.Keys.OrderBy(k => k).ToList();

        // Shared output types.
        var financialCostType = new ObjectGraphType { Name = "FinancialCost" };
        financialCostType.Field<FloatGraphType>("amount");
        financialCostType.Field<StringGraphType>("currency");

        var costActualType = new ObjectGraphType { Name = "CostActual" };
        costActualType.AddField(new FieldType
        {
            Name = "financial",
            ResolvedType = financialCostType,
        });
        costActualType.Field<StringGraphType>("varianceFromEstimate");

        var resolutionType = new ObjectGraphType { Name = "Resolution" };
        resolutionType.Field<NonNullGraphType<StringGraphType>>("action");
        resolutionType.Field<NonNullGraphType<StringGraphType>>("recoveryClass");
        resolutionType.Field<StringGraphType>("requires");
        resolutionType.Field<StringGraphType>("grantableBy");

        var anipFailureType = new ObjectGraphType { Name = "ANIPFailure" };
        anipFailureType.Field<NonNullGraphType<StringGraphType>>("type");
        anipFailureType.Field<NonNullGraphType<StringGraphType>>("detail");
        anipFailureType.AddField(new FieldType
        {
            Name = "resolution",
            ResolvedType = resolutionType,
        });
        anipFailureType.Field<NonNullGraphType<BooleanGraphType>>("retry");

        var queryType = new ObjectGraphType { Name = "Query" };
        var mutationType = new ObjectGraphType { Name = "Mutation" };

        foreach (var name in capNames)
        {
            var decl = service.GetCapabilityDeclaration(name);
            if (decl == null) continue;

            var pascal = GraphQLResponseMapper.ToPascalCase(name);
            var camel = GraphQLResponseMapper.ToCamelCase(name);

            // Result type for this capability.
            var resultType = new ObjectGraphType { Name = pascal + "Result" };
            resultType.Field<NonNullGraphType<BooleanGraphType>>("success");
            resultType.AddField(new FieldType
            {
                Name = "result",
                ResolvedType = new JsonGraphType(),
            });
            resultType.AddField(new FieldType
            {
                Name = "costActual",
                ResolvedType = costActualType,
            });
            resultType.AddField(new FieldType
            {
                Name = "failure",
                ResolvedType = anipFailureType,
            });

            // Build the field.
            var field = new FieldType
            {
                Name = camel,
                ResolvedType = new NonNullGraphType(resultType),
                Resolver = new AnipFieldResolver(service, name),
            };

            // Build arguments from inputs.
            if (decl.Inputs != null)
            {
                var queryArgs = new QueryArguments();
                foreach (var inp in decl.Inputs)
                {
                    var argName = GraphQLResponseMapper.ToCamelCase(inp.Name);
                    var baseType = AnipTypeToGraphQL(inp.Type);
                    IGraphType argType = inp.Required
                        ? new NonNullGraphType(baseType)
                        : baseType;

                    queryArgs.Add(new QueryArgument(argType) { Name = argName });
                }
                field.Arguments = queryArgs;
            }

            var seType = decl.SideEffect?.Type ?? "";
            if (string.IsNullOrEmpty(seType) || seType == "read")
            {
                queryType.AddField(field);
            }
            else
            {
                mutationType.AddField(field);
            }
        }

        // graphql-dotnet requires a Query type; add a dummy if empty.
        if (queryType.Fields.Count() == 0)
        {
            queryType.AddField(new FieldType
            {
                Name = "_empty",
                ResolvedType = new StringGraphType(),
            });
        }

        var schema = new Schema
        {
            Query = queryType,
        };

        if (mutationType.Fields.Count() > 0)
        {
            schema.Mutation = mutationType;
        }

        var sdl = GenerateSDL(service, capNames);

        return new SchemaResult(schema, sdl);
    }

    /// <summary>
    /// Generates SDL text including custom @anip* directives.
    /// </summary>
    private static string GenerateSDL(AnipService service, List<string> capNames)
    {
        var lines = new List<string>
        {
            "directive @anipSideEffect(type: String!, rollbackWindow: String) on FIELD_DEFINITION",
            "directive @anipCost(certainty: String!, currency: String, rangeMin: Float, rangeMax: Float) on FIELD_DEFINITION",
            "directive @anipRequires(capabilities: [String!]!) on FIELD_DEFINITION",
            "directive @anipScope(scopes: [String!]!) on FIELD_DEFINITION",
            "",
            "scalar JSON",
            "",
            "type CostActual { financial: FinancialCost, varianceFromEstimate: String }",
            "type FinancialCost { amount: Float, currency: String }",
            "type ANIPFailure { type: String!, detail: String!, resolution: Resolution, retry: Boolean! }",
            "type Resolution { action: String!, recoveryClass: String!, requires: String, grantableBy: String }",
            "type RestrictedCapability { capability: String!, reason: String!, reasonType: String!, grantableBy: String!, unmetTokenRequirements: [String!]!, resolutionHint: String }",
            "type DeniedCapability { capability: String!, reason: String!, reasonType: String! }",
            "",
        };

        var queries = new List<string>();
        var mutations = new List<string>();

        foreach (var name in capNames)
        {
            var decl = service.GetCapabilityDeclaration(name);
            if (decl == null) continue;

            var pascal = GraphQLResponseMapper.ToPascalCase(name);
            var camel = GraphQLResponseMapper.ToCamelCase(name);

            // Result type.
            lines.Add($"type {pascal}Result {{ success: Boolean!, result: JSON, costActual: CostActual, failure: ANIPFailure }}");

            var argsStr = BuildSDLArgs(decl);
            var dirStr = BuildSDLDirectives(decl);

            var fieldLine = $"  {camel}{argsStr}: {pascal}Result! {dirStr}";

            var seType = decl.SideEffect?.Type ?? "";
            if (string.IsNullOrEmpty(seType) || seType == "read")
            {
                queries.Add(fieldLine);
            }
            else
            {
                mutations.Add(fieldLine);
            }
        }

        lines.Add("");
        if (queries.Count > 0)
        {
            lines.Add("type Query {");
            lines.AddRange(queries);
            lines.Add("}");
        }
        if (mutations.Count > 0)
        {
            lines.Add("type Mutation {");
            lines.AddRange(mutations);
            lines.Add("}");
        }

        return string.Join("\n", lines);
    }

    private static string BuildSDLArgs(CapabilityDeclaration decl)
    {
        if (decl.Inputs == null || decl.Inputs.Count == 0)
        {
            return "";
        }

        var args = new List<string>();
        foreach (var inp in decl.Inputs)
        {
            var argName = GraphQLResponseMapper.ToCamelCase(inp.Name);
            var gqlType = AnipTypeToGraphQLName(inp.Type);
            if (inp.Required)
            {
                gqlType += "!";
            }
            args.Add($"{argName}: {gqlType}");
        }
        return "(" + string.Join(", ", args) + ")";
    }

    private static string BuildSDLDirectives(CapabilityDeclaration decl)
    {
        var parts = new List<string>();

        // @anipSideEffect
        var seType = decl.SideEffect?.Type ?? "";
        if (string.IsNullOrEmpty(seType)) seType = "read";
        var seDir = new StringBuilder($"@anipSideEffect(type: \"{seType}\"");
        if (!string.IsNullOrEmpty(decl.SideEffect?.RollbackWindow))
        {
            seDir.Append($", rollbackWindow: \"{decl.SideEffect.RollbackWindow}\"");
        }
        seDir.Append(')');
        parts.Add(seDir.ToString());

        // @anipCost
        if (decl.Cost != null)
        {
            var certainty = decl.Cost.Certainty;
            if (string.IsNullOrEmpty(certainty)) certainty = "estimate";
            var costDir = new StringBuilder($"@anipCost(certainty: \"{certainty}\"");
            if (decl.Cost.Financial != null)
            {
                if (!string.IsNullOrEmpty(decl.Cost.Financial.Currency))
                {
                    costDir.Append($", currency: \"{decl.Cost.Financial.Currency}\"");
                }
                if (decl.Cost.Financial.RangeMin != null)
                {
                    costDir.Append($", rangeMin: {decl.Cost.Financial.RangeMin}");
                }
                if (decl.Cost.Financial.RangeMax != null)
                {
                    costDir.Append($", rangeMax: {decl.Cost.Financial.RangeMax}");
                }
            }
            costDir.Append(')');
            parts.Add(costDir.ToString());
        }

        // @anipRequires
        if (decl.Requires != null && decl.Requires.Count > 0)
        {
            var capStrs = decl.Requires.Select(r => $"\"{r.Capability}\"");
            parts.Add($"@anipRequires(capabilities: [{string.Join(", ", capStrs)}])");
        }

        // @anipScope
        if (decl.MinimumScope != null && decl.MinimumScope.Count > 0)
        {
            var scopeVals = decl.MinimumScope.Select(s => $"\"{s}\"");
            parts.Add($"@anipScope(scopes: [{string.Join(", ", scopeVals)}])");
        }

        return string.Join(" ", parts);
    }

    /// <summary>
    /// Result of building the schema: executable schema + SDL text.
    /// </summary>
    public record SchemaResult(ISchema Schema, string Sdl);
}

/// <summary>
/// A simple JSON scalar type for graphql-dotnet.
/// Passes through any object value without conversion.
/// </summary>
internal class JsonGraphType : ScalarGraphType
{
    public JsonGraphType()
    {
        Name = "JSON";
    }

    public override object? ParseLiteral(GraphQLParser.AST.GraphQLValue value) => value;
    public override object? ParseValue(object? value) => value;
    public override object? Serialize(object? value) => value;
}

/// <summary>
/// Data resolver for ANIP capability fields. Handles auth and invocation.
/// </summary>
internal class AnipFieldResolver : global::GraphQL.Resolvers.IFieldResolver
{
    private readonly AnipService _service;
    private readonly string _capName;

    public AnipFieldResolver(AnipService service, string capName)
    {
        _service = service;
        _capName = capName;
    }

    public ValueTask<object?> ResolveAsync(IResolveFieldContext context)
    {
        // Extract auth header from context.
        var authHeader = context.GetArgument<string>("__authHeader");
        // Also check user context.
        if (string.IsNullOrEmpty(authHeader))
        {
            authHeader = context.UserContext.TryGetValue("authHeader", out var ah) ? ah as string : null;
        }

        var bearer = ExtractBearer(authHeader);

        if (string.IsNullOrEmpty(bearer))
        {
            var noAuthResult = GraphQLResponseMapper.BuildGraphQLResponse(new Dictionary<string, object?>
            {
                ["success"] = false,
                ["failure"] = new Dictionary<string, object?>
                {
                    ["type"] = Constants.FailureAuthRequired,
                    ["detail"] = "Authorization header required",
                    ["resolution"] = new Dictionary<string, object?> { ["action"] = "provide_credentials", ["recovery_class"] = Constants.RecoveryClassForAction("provide_credentials") },
                    ["retry"] = true,
                },
            });
            return new ValueTask<object?>(noAuthResult);
        }

        DelegationToken token;
        try
        {
            token = GraphQLAuthBridge.ResolveAuth(bearer, _service, _capName);
        }
        catch (AnipError e)
        {
            var failure = new Dictionary<string, object?>
            {
                ["type"] = e.ErrorType,
                ["detail"] = e.Detail,
                ["retry"] = e.Retry,
            };
            if (e.Resolution != null)
            {
                failure["resolution"] = new Dictionary<string, object?>
                {
                    ["action"] = e.Resolution.Action,
                    ["recovery_class"] = !string.IsNullOrEmpty(e.Resolution.RecoveryClass)
                        ? e.Resolution.RecoveryClass
                        : Constants.RecoveryClassForAction(e.Resolution.Action),
                };
            }
            var authFailResult = GraphQLResponseMapper.BuildGraphQLResponse(new Dictionary<string, object?>
            {
                ["success"] = false,
                ["failure"] = failure,
            });
            return new ValueTask<object?>(authFailResult);
        }
        catch (Exception)
        {
            var internalFailResult = GraphQLResponseMapper.BuildGraphQLResponse(new Dictionary<string, object?>
            {
                ["success"] = false,
                ["failure"] = new Dictionary<string, object?>
                {
                    ["type"] = Constants.FailureInternalError,
                    ["detail"] = "Authentication failed",
                    ["retry"] = false,
                },
            });
            return new ValueTask<object?>(internalFailResult);
        }

        // Convert camelCase args to snake_case for ANIP invocation.
        var snakeArgs = new Dictionary<string, object?>();
        if (context.Arguments != null)
        {
            foreach (var arg in context.Arguments)
            {
                snakeArgs[GraphQLResponseMapper.ToSnakeCase(arg.Key)] = arg.Value.Value;
            }
        }

        var result = _service.Invoke(_capName, token, snakeArgs, new InvokeOpts());

        var response = GraphQLResponseMapper.BuildGraphQLResponse(result);
        return new ValueTask<object?>(response);
    }

    private static string? ExtractBearer(string? authHeader)
    {
        if (authHeader != null && authHeader.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
        {
            return authHeader["Bearer ".Length..].Trim();
        }
        return null;
    }
}
